### ISSUE TITLE
Potential circular import in `prefetcher`

### ISSUE THREAD

**Issue by ejguan** [pytorch/data#931]
### 🐛 Describe the bug

In prefetcher.py, `dataloader2` is imported. There is a potential circular import issue if `dataloader2` needs to take some dependency on DataPipe.
https://github.com/pytorch/data/blob/fbee6f75c9e630ea793116caea58911d5ad7d6e0/torchdata/datapipes/iter/util/prefetcher.py#L13

We need to guarantee the dependency flow is `dataloader2` depends on `datapipes` but not vice versa.

### Versions

main

**Comment by NivekT** [comment]
I think one solution may be to move those exceptions from `torchdata.dataloader2` to `torchdata.datapipes`, because they are meant to be raised by DataPipes anyway. They are currently only raised by `_IterateQueueDataPipes`.

```python
try:
    ...
except communication.iter.InvalidStateResetRequired:
    prefetch_data.stop_iteration = True
except communication.iter.TerminateRequired:
    prefetch_data.run_prefetcher = False
```
Namely, these ones:
```
InvalidStateResetRequired
TerminateRequired
```

**Comment by ejguan** [comment]
Those two DataPipes are only used by DataLoader2, right? It might be better to put them somewhere in `torchdata/dataloader2`

**Comment by NivekT** [comment]
I think the issue here is that `prefetcher.py` is importing those exceptions to catch them?

**Comment by ejguan** [comment]
TBH, I am not sure if exposing `Prefetcher` to users or not. Like what I did in https://github.com/pytorch/data/pull/919 for `_IterateQueueDataPipes`, I just moved it under `dataloader2` directory to prevent circular import as it's not a user-facing DataPipe.
I am not sure about `Prefetcher`, though.

**Comment by NivekT** [comment]
I think exposing `Prefetcher` to users is necessary for prefetching from source.

**Comment by ejguan** [comment]
If that's the case, I think we need to move communication module out of `dataloader2`. Then, the dependency graph becomes,
```mermaid
graph TD;
    communication-->datapipes;
    datapipes-->dataloader2;
    communication-->dataloader2;
```

Another solution might be figure out why those communication exceptions has to be placed in `prefetch` and can we move it out.

**Comment by ejguan** [comment]
I am super confused by `InvalidStateResetRequired` in https://github.com/pytorch/data/blob/fbee6f75c9e630ea793116caea58911d5ad7d6e0/torchdata/datapipes/iter/util/prefetcher.py#L72-L73

The only place raising `InvalidStateResetRequired` is from [`_IterateQueueDataPipes `](https://github.com/pytorch/data/blob/fbee6f75c9e630ea793116caea58911d5ad7d6e0/torchdata/dataloader2/reading_service.py#L143-L144), and it is raised when `InvalidStateResponse` is returned from queue. And, the `response_invalid_state ` method attached to the protocol server in worker process needs to be called. However, `response_invalid_state` is only called when another `InvalidStateResetRequired` is raised in https://github.com/pytorch/data/blob/fbee6f75c9e630ea793116caea58911d5ad7d6e0/torchdata/dataloader2/communication/iter.py#L159-L160

The whole logic for `InvalidStateResetRequired` is self referenced and never reached. I am not sure why we still need. Can you please help me to see if I understand this exception flow correctly.

**Comment by wenleix** [comment]
noob question: seems Exception is all you need from those packages. Does it make sense to refactor these two exceptions into some module like `common.exception`?

**Comment by ejguan** [comment]
@wenleix It totally make sense to me.
I just don't fully understand what was the original reason to introduce them. Let me first figure it out and see what would be the best way to refactor the structure.

**Comment by ejguan** [comment]
Finally, I found the first PR that introduced `InvalidStateResetRequired` in https://github.com/facebookexternal/torchdata/pull/21

This state is introduced when not all `QueueWrapper` in the main process have requested to reset epoch. This problem is prevented by `_IterateQueueDataPipes`, which wraps all `QeueuWrappers` for all work processes and always reset all of them in https://github.com/pytorch/data/blob/fbee6f75c9e630ea793116caea58911d5ad7d6e0/torchdata/dataloader2/reading_service.py#L159-L165

So, I think `InvalidStateResetRequired` can be removed.


It seems `TerminateRequired` is still needed for `prefetch` to early exit prefetch thread without spending more time on fetching data from prior DataPipe

**Comment by NivekT** [comment]
Fixed by #1014
